### PR TITLE
vm_map_protect: imply VM_PROT_CAP from max_protection

### PR DESCRIPTION
**Pull request by brooksdavis**
Always check entry->max_protection for VM_PROT_CAP when deciding if we should imply VM_PROT_CAP for readable and writable pages.

Previously we would only check it if we were setting a new max_protection (which mprotect generally didn't do) which meant we didn't add VM_PROT_CAP when the current permission were VM_PROT_NONE.

Fixes: #1818

### CODE REVIEW

**Comment by brooksdavis** [comment]
The one thing I'm not completely happy with here is that I'd rather not be implying capability protections unless they were implied on the reservation at initial mapping. Right now, if they are there, they were implied (or it's a mapping created by the kernel where you could argue they were implied if we also provide a VMEM-bearing capability). That could change in the future when we finally extend mmap to support explicit capability permissions at which point I think we'll want to add a flag in the entry to tell if permissions were implied or not.

File: bin/cheribsdtest/cheribsdtest_cheriabi.c [hunk 1]
```diff
@@ -284,6 +284,22 @@ CHERIBSDTEST(cheriabi_munmap_invalid_ptr,
 	cheribsdtest_success();
 }
 
+CHERIBSDTEST(cheriabi_mprotect_upgrade_prot,
+    "Check that upgrading from PROT_NONE works")
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_cheriabi.c:288]: Do you want to mention "cap" or capability in the description?  Maybe "Check that upgrading from PROT_NONE includes capability permissions"?

File: bin/cheribsdtest/cheribsdtest_cheriabi.c [hunk 2]
```diff
@@ -284,6 +284,22 @@ CHERIBSDTEST(cheriabi_munmap_invalid_ptr,
...
+{
+	void * volatile *p;
+
+	p = CHERIBSDTEST_CHECK_SYSCALL(mmap(NULL, PAGE_SIZE,
+	    PROT_NONE | PROT_MAX(PROT_READ | PROT_WRITE),
+	    MAP_ANON | MAP_PRIVATE, -1, 0));
+	CHERIBSDTEST_CHECK_SYSCALL(mprotect(__DEVOLATILE(void *, p), PAGE_SIZE,
+	    PROT_READ | PROT_WRITE));
+	/* Attempt to store a capability */
```
Comment (bsdjhb) [inline bin/cheribsdtest/cheribsdtest_cheriabi.c:297]: Nit: blank line before comment.

File: sys/vm/vm_map.c [hunk 3]
```diff
@@ -3326,7 +3326,7 @@ vm_map_protect(vm_map_t map, vm_offset_t start, vm_offset_t end,
 
 		/*
 		 * When VM_MAP_PROTECT_KEEP_CAP is used (for mprotect()),
```
Comment (bsdjhb) [inline sys/vm/vm_map.c:3328]: Suggestions for commit log:  s/mprotect generally didn't do/mprotect generally doesn't do/ and s/current permission were/current permissions were/

File: sys/vm/vm_map.c [hunk 4]
```diff
@@ -3337,9 +3337,8 @@ vm_map_protect(vm_map_t map, vm_offset_t start, vm_offset_t end,
 		 * potential compatibility breakage.  The alternative
 		 * would maximize compatibility.
 		 */
-		if (flags & VM_MAP_PROTECT_KEEP_CAP &&
-		    (((flags & VM_MAP_PROTECT_SET_MAXPROT) != 0 && (entry->max_protection & VM_PROT_CAP) != 0) ||
-		    ((flags & VM_MAP_PROTECT_SET_MAXPROT) == 0 && (entry->protection & VM_PROT_CAP) != 0))) {
+		if ((flags & VM_MAP_PROTECT_KEEP_CAP) != 0 &&
```
Comment (bsdjhb) [inline sys/vm/vm_map.c:3340]: I believe this is ok to only check max_protection.  In particular, entry->protection can only be a subset, so there's no case where entry->protection would contain some VM_PROT_CAP bits and entry->max_protection would not.

File: sys/vm/vm_map.c [hunk 5]
```diff
@@ -3340,6 +3340,7 @@ vm_map_protect(vm_map_t map, vm_offset_t start, vm_offset_t end,
 		if ((flags & VM_MAP_PROTECT_KEEP_CAP) != 0 &&
 		    (entry->max_protection & VM_PROT_CAP) != 0) {
 			new_prot = VM_PROT_ADD_CAP(new_prot);
+			new_maxprot = VM_PROT_ADD_CAP(new_maxprot);
```
Comment (bsdjhb) [inline sys/vm/vm_map.c:3343]: Do we want a test for this?  Sounds like you could do a RW mmap, store a capability, downgrade to read-only and then see if you can read the tagged capability?

**Review by bsdjhb — APPROVED** [review]
Generally LGTM

File: sys/vm/vm_map.c [hunk 6]
```diff
@@ -3337,9 +3337,8 @@ vm_map_protect(vm_map_t map, vm_offset_t start, vm_offset_t end,
 		 * potential compatibility breakage.  The alternative
 		 * would maximize compatibility.
 		 */
-		if (flags & VM_MAP_PROTECT_KEEP_CAP &&
-		    (((flags & VM_MAP_PROTECT_SET_MAXPROT) != 0 && (entry->max_protection & VM_PROT_CAP) != 0) ||
-		    ((flags & VM_MAP_PROTECT_SET_MAXPROT) == 0 && (entry->protection & VM_PROT_CAP) != 0))) {
+		if ((flags & VM_MAP_PROTECT_KEEP_CAP) != 0 &&
+		    (entry->max_protection & VM_PROT_CAP) != 0) {
```
Comment (markjdb) [inline sys/vm/vm_map.c:3341]: I don't think it affects correctness, but shouldn't we skip this block if `(flags & VM_MAP_PROTECT_SET_PROT) == 0`?

File: sys/vm/vm_map.c [hunk 7]
```diff
@@ -3326,7 +3326,7 @@ vm_map_protect(vm_map_t map, vm_offset_t start, vm_offset_t end,
 
 		/*
 		 * When VM_MAP_PROTECT_KEEP_CAP is used (for mprotect()),
-		 * upgrade* new_prot to include the associated VM_PROT_CAP
+		 * upgrade new_prot to include the associated VM_PROT_CAP
 		 * bits and retry the the scan.  This means that the
 		 * mprotect will fail if it spans map entries with and
 		 * without VM_PROT_CAP set.
```
Comment (markjdb) [inline sys/vm/vm_map.c:3332]: I'd clarify this comment a bit: "... with and without VM_PROT_CAP set in max_protection."

File: sys/vm/vm_map.c (hunk 5 above)
Comment (brooksdavis) [reply sys/vm/vm_map.c:3346]: I've added a test to do this and verify that store is forbidden.

File: sys/vm/vm_map.c (hunk 6 above)
Comment (brooksdavis) [reply sys/vm/vm_map.c:3344]: I added an assert to accomplish this as it's a kernel bug to call with VM_MAP_PROTECT_KEEP_CAP and not at least one of VM_MAP_PROTECT_SET_PROT or VM_MAP_PROTECT_SET_MAXPROT

**Review by bsdjhb — APPROVED** [review]

**Pull request merged**